File: 03. Stack and Queue/04. Maximal Rectangle/main.js
function largestRectangleArea(heights) {
    let stack = [];
    let maxArea = 0;
    heights.push(0); // Sentinel value

    for (let i = 0; i < heights.length; i++) {
        while (stack.length && heights[i] < heights[stack[stack.length - 1]]) {
            let height = heights[stack.pop()];
            let width = stack.length ? i - stack[stack.length - 1] - 1 : i;
            maxArea = Math.max(maxArea, height * width);
        }
        stack.push(i);
    }

    return maxArea;
}

function maximalRectangle(matrix) {
    if (!matrix.length) return 0;
    let cols = matrix[0].length;
    let heights = Array(cols).fill(0);
    let maxArea = 0;

    for (let row of matrix) {
        for (let j = 0; j < cols; j++) {
            heights[j] = row[j] === '1' ? heights[j] + 1 : 0;
        }
        maxArea = Math.max(maxArea, largestRectangleArea(heights));
    }

    return maxArea;
}

// Main function
function main() {
    let matrix = [
        ['1', '0', '1', '0', '0'],
        ['1', '0', '1', '1', '1'],
        ['1', '1', '1', '1', '1'],
        ['1', '0', '0', '1', '0']
    ];

    console.log("Maximal Rectangle Area:", maximalRectangle(matrix));
}

// Run main function
main();

File: 03. Stack and Queue/04. Maximal Rectangle/main.py
def largestRectangleArea(heights):
    stack = []
    max_area = 0
    heights.append(0)  # Sentinel value

    for i in range(len(heights)):
        while stack and heights[i] < heights[stack[-1]]:
            height = heights[stack.pop()]
            width = i if not stack else i - stack[-1] - 1
            max_area = max(max_area, height * width)
        stack.append(i)

    return max_area

def maximalRectangle(matrix):
    if not matrix:
        return 0

    cols = len(matrix[0])
    heights = [0] * cols
    max_area = 0

    for row in matrix:
        for j in range(cols):
            heights[j] = heights[j] + 1 if row[j] == '1' else 0
        max_area = max(max_area, largestRectangleArea(heights))

    return max_area

# Main function
def main():
    matrix = [
        ['1', '0', '1', '0', '0'],
        ['1', '0', '1', '1', '1'],
        ['1', '1', '1', '1', '1'],
        ['1', '0', '0', '1', '0']
    ]

    print("Maximal Rectangle Area:", maximalRectangle(matrix))

# Run main function
if __name__ == "__main__":
    main()
